File: metal/paddle-mobile/paddle-mobile/Operators/Kernels/PoolKernel.swift
/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 
 http://www.apache.org/licenses/LICENSE-2.0
 
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License. */

import Foundation

struct PoolMetalParam {
  let ksizeX: Int32
  let ksizeY: Int32
  let strideX: Int32
  let strideY: Int32
  let paddingX: Int32
  let paddingY: Int32
  let poolType: Int32
}

class PoolKernel<P: PrecisionType>: Kernel, Computable{
  
  required init(device: MTLDevice, param: PoolParam<P>) {
    super.init(device: device, inFunctionName: "pool")
    param.output.initTexture(device: device)
  }
  
  func compute(commandBuffer: MTLCommandBuffer, param: PoolParam<P>) throws {
    guard let encoder = commandBuffer.makeComputeCommandEncoder() else {
      throw PaddleMobileError.predictError(message: " encoder is nil")
    }
    encoder.setTexture(param.input.metalTexture, index: 0)
    encoder.setTexture(param.output.metalTexture, index: 1)
    var poolType: Int32
    switch param.poolType {
    case "max":
      poolType = 0
    case "avg":
      poolType = 1
    default:
      throw PaddleMobileError.predictError(message: " unknown pooltype " + param.poolType)
    }
    var pmp = PoolMetalParam.init(
      ksizeX: param.ksize[0],
      ksizeY: param.ksize[1],
      strideX: param.stride[0],
      strideY: param.stride[1],
      paddingX: param.padding[0],
      paddingY: param.padding[1],
      poolType: poolType
    )
    encoder.setBytes(&pmp, length: MemoryLayout<PoolMetalParam>.size, index: 0)
    encoder.dispatch(computePipline: pipline, outTexture: param.output.metalTexture)
    encoder.endEncoding()
  }
}

File: src/operators/kernel/fpga/KD/pool_kernel.cpp
/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */
#ifdef POOL_OP

#include "operators/kernel/pool_kernel.h"
#include "fpga/KD/pes/pooling_pe.hpp"

class PoolingArgs;
namespace paddle_mobile {
namespace operators {

template <>
bool PoolKernel<FPGA, float>::Init(PoolParam<FPGA>* param) {
  param->Output()->mutable_data<half>();

  zynqmp::PoolingPE& pe = param->context().pe<zynqmp::PoolingPE>();
  zynqmp::PoolingParam& pool_param = pe.param();

  pool_param.input = param->Input()->zynqmpTensor();
  pool_param.output = param->Output()->zynqmpTensor();
  pool_param.type = param->PoolingType() == "max"
                        ? zynqmp::PoolingType::MAX
                        : zynqmp::PoolingType::AVERAGE;
  pool_param.globalPooling = param->isGlobalPooling();
  pool_param.kernelSize = param->Ksize();
  pool_param.strides = param->Strides();
  pool_param.paddings = param->Paddings();

  pe.init();
  pe.apply();
  return true;
}

template <>
void PoolKernel<FPGA, float>::Compute(const PoolParam<FPGA>& param) {
  zynqmp::Context& context = const_cast<zynqmp::Context&>(param.context_);
  zynqmp::PoolingPE& pe = context.pe<zynqmp::PoolingPE>();
  pe.dispatch();

  // param.Output()->zynqmpTensor()->printScale();
}
}  // namespace operators
}  // namespace paddle_mobile

#endif
